File: ios/CellularChat/Ranging/RangingCoordinator.swift
import Foundation
import Combine
import CellularChatCore

/// Bounded exponential backoff (PROTOCOL_V2.md §12): initial 5 s, ×2, cap 60 s.
struct BoundedBackoff {
    let initial: TimeInterval
    let factor: Double
    let cap: TimeInterval
    private var current: TimeInterval?

    init(initial: TimeInterval = 5, factor: Double = 2, cap: TimeInterval = 60) {
        self.initial = initial
        self.factor = factor
        self.cap = cap
    }

    mutating func next() -> TimeInterval {
        let value = current.map { min($0 * factor, cap) } ?? initial
        current = value
        return value
    }

    mutating func reset() { current = nil }
}

/// Ranging method selection + degradation (PROTOCOL_V2.md §12). Selects the
/// method deterministically from both CapabilitySets, drives the platform
/// ranger, and on UWB invalidation/sample loss keeps the authenticated session,
/// falls back immediately to BLE RSSI proximity, and retries UWB with bounded
/// backoff. A ranging failure is never surfaced as an auth/pairing failure.
@MainActor
final class RangingCoordinator: ObservableObject {

    @Published private(set) var measurement: Measurement?
    @Published private(set) var selection: RangingSelection?
    @Published private(set) var stateText: String = "찾는 중"

    /// Sends an authenticated ranging session message (§8); wired by the session
    /// layer. The coordinator owns the monotonic `attemptId` and stamps it into
    /// every ranging body (§10/§12).
    var sendMessage: ((SessionMsgType, CBOR) -> Void)?

    private let peerRanger = ApplePeerRanger()
    private let interopRanger = AndroidInteropRanger()
    private let rssiFilter = RSSIProximityFilter()
    private var backoff = BoundedBackoff()
    private var retryTask: Task<Void, Never>?
    private var active = false

    // Ranging-attempt negotiation state (§8 offer/accept/start, §10/§12).
    // `attemptCounter` yields a monotonically increasing attemptId per sid;
    // `startedAttempt` gives (sid, attemptId) idempotency so a duplicate
    // start/stop is a no-op. `isOfferer` is the ni_peer ranging controller.
    private var attemptCounter: UInt64 = 0
    private var currentAttemptId: UInt64?
    private var startedAttempt: UInt64?
    private var isOfferer = false

    init() {
        peerRanger.onSendToken = { [weak self] data in self?.sendRangingData(.niToken, data) }
        peerRanger.onMeasurement = { [weak self] m in self?.handleUWB(measurement: m) }
        interopRanger.onSendShareable = { [weak self] data in self?.sendRangingData(.appleShareable, data) }
        interopRanger.onMeasurement = { [weak self] m in self?.handleUWB(measurement: m) }
    }

    private func nextAttemptId() -> UInt64 { attemptCounter += 1; return attemptCounter }

    private func sendMap(_ type: SessionMsgType, _ pairs: [CBORPair]) { sendMessage?(type, .map(pairs)) }

    /// Sends `ni_token`/`apple_shareable` for the current attempt (§10/§12).
    private func sendRangingData(_ type: SessionMsgType, _ data: Data) {
        guard let id = currentAttemptId else { return }
        sendMap(type, [CBORPair(.uint(1), .uint(id)), CBORPair(.uint(2), .bytes(Array(data)))])
    }

    // MARK: lifecycle

    func start(local: CapabilitySet, peer: CapabilitySet, isInitiator: Bool) {
        let selection = RangingSelector.select(local: local, peer: peer)
        self.selection = selection
        active = true
        backoff.reset()
        attemptCounter = 0
        currentAttemptId = nil
        startedAttempt = nil
        // The ni_peer ranging controller (offerer) is the Noise initiator; every
        // other method is peer-driven (Android sends apple_config directly, never
        // an offer) or a local fallback, so iOS offers only for ni_peer.
        isOfferer = (selection.method == .niPeer) && isInitiator
        beginAttempt()
    }

    func stop() {
        active = false
        retryTask?.cancel()
        retryTask = nil
        peerRanger.stop()
        interopRanger.stop()
        rssiFilter.reset()
        measurement = nil
        stateText = "중지됨"
    }

    /// Starts one ranging attempt with a fresh `attemptId` (§12). The offerer
    /// drives ranging_offer → ranging_start; the controlee waits for the peer's
    /// offer (ni_peer) or apple_config (uwb_apple_interop).
    private func beginAttempt() {
        guard active, let selection else { return }
        switch selection.method {
        case .niPeer:
            if isOfferer {
                let id = nextAttemptId()
                currentAttemptId = id
                stateText = "UWB 측정 협상 중"
                sendMap(.rangingOffer, [CBORPair(.uint(1), .uint(id)),
                                        CBORPair(.uint(2), .uint(RangingMethod.niPeer.rawValue))])
            } else {
                stateText = "UWB 방향/거리 측정 준비 중"   // await ranging_offer
            }
        case .uwbAppleInterop:
            stateText = "UWB 연동 측정 준비 중 (실기기 검증 필요)"   // await apple_config
        case .uwbAndroidOob:
            fallbackToRSSI()   // Android-to-Android only; degrade on an iOS device.
        case .bleRssi:
            stateText = "근접도만 측정 (BLE)"   // driven by feedRSSI(_:)
        }
    }

    // MARK: inbound ranging messages (from the authenticated session)

    /// Routes an inbound ranging session message (§8), mirroring the Android
    /// peer's `onSessionMessage`. Unknown fields and duplicate operations are
    /// safe no-ops (§10 (sid, attemptId) idempotency).
    func handleSessionMessage(_ type: SessionMsgType, body: CBOR) {
        guard active else { return }
        let attemptId = body.value(forKey: 1)?.asUInt
        switch type {
        case .rangingOffer:
            guard let id = attemptId,
                  let raw = body.value(forKey: 2)?.asUInt,
                  let method = RangingMethod(rawValue: raw) else { return }
            onRangingOffer(attemptId: id, method: method)
        case .rangingAccept:
            if let id = attemptId { onRangingAccept(attemptId: id) }
        case .rangingStart:
            if let id = attemptId { onRangingStart(attemptId: id) }
        case .rangingStop:
            if let id = attemptId { onRangingStop(attemptId: id) }
        case .rangingError:
            if let id = attemptId { onRangingError(attemptId: id) }
        case .niToken:
            if let id = attemptId, let data = body.value(forKey: 2)?.asBytes {
                onNiToken(attemptId: id, data: Data(data))
            }
        case .appleConfig:
            if let id = attemptId, let data = body.value(forKey: 2)?.asBytes {
                onAppleConfig(attemptId: id, data: Data(data))
            }
        default:
            break   // apple_shareable / oob_data are outbound-only on iOS
        }
    }

    // Controlee: the peer offered ranging → accept and await ranging_start (§8).
    private func onRangingOffer(attemptId: UInt64, method: RangingMethod) {
        currentAttemptId = attemptId
        sendMap(.rangingAccept, [CBORPair(.uint(1), .uint(attemptId)),
                                 CBORPair(.uint(2), .uint(method.rawValue))])
    }

    // Offerer: the peer accepted → send ranging_start and begin the method.
    private func onRangingAccept(attemptId: UInt64) {
        guard attemptId == currentAttemptId else { return }
        sendMap(.rangingStart, [CBORPair(.uint(1), .uint(attemptId))])
        beginMethod(attemptId: attemptId)
    }

    // Controlee: the offerer started the attempt → begin the method.
    private func onRangingStart(attemptId: UInt64) {
        guard attemptId == currentAttemptId else { return }
        beginMethod(attemptId: attemptId)
    }

    // A duplicate (sid, attemptId) start is an idempotent no-op (§10).
    private func beginMethod(attemptId: UInt64) {
        guard active, startedAttempt != attemptId else { return }
        startedAttempt = attemptId
        if selection?.method == .niPeer { peerRanger.start(enableEDM: selection?.edm ?? false) }
    }

    private func onNiToken(attemptId: UInt64, data: Data) {
        guard attemptId == currentAttemptId else { return }
        peerRanger.receivePeerToken(data)
    }

    // Android drives uwb_apple_interop by sending apple_config directly (no
    // offer); adopt its attemptId so apple_shareable echoes it (§10/§12).
    private func onAppleConfig(attemptId: UInt64, data: Data) {
        guard startedAttempt != attemptId else { return }   // idempotent
        currentAttemptId = attemptId
        startedAttempt = attemptId
        interopRanger.start(appleConfig: data)
    }

    // Peer stopped this attempt: keep the session, drop the ranger (§10/§12).
    private func onRangingStop(attemptId: UInt64) {
        guard startedAttempt == attemptId else { return }   // idempotent / stale
        startedAttempt = nil
        switch selection?.method {
        case .niPeer?: peerRanger.stop()
        case .uwbAppleInterop?: interopRanger.stop()
        default: break
        }
    }

    // Peer reported a ranging error: never an auth failure — degrade + retry (§12).
    private func onRangingError(attemptId: UInt64) {
        guard attemptId == currentAttemptId else { return }
        startedAttempt = nil
        fallbackToRSSI()
        scheduleUWBRetry()
    }

    /// Feed a raw RSSI reading from the BLE link for the proximity fallback (§12).
    func feedRSSI(_ rssi: Double) {
        guard active else { return }
        let band = rssiFilter.add(rssi: rssi)
        // Only surface RSSI proximity when UWB is not currently providing a sample.
        if selection?.method == .bleRssi || measurement == nil || measurement?.proximity != nil {
            measurement = Measurement(timestamp: Date(), method: .bleRssi,
                                      distanceMeters: nil, horizontalAngleRadians: nil, proximity: band)
        }
    }

    // MARK: UWB result handling + backoff

    private func handleUWB(measurement m: Measurement?) {
        guard active else { return }
        if let m {
            backoff.reset()
            measurement = m
            stateText = m.horizontalAngleRadians != nil ? "방향 측정 중"
                : (m.distanceMeters != nil ? "거리 측정 중" : "신호 탐색 중")
        } else {
            // UWB sample lost/invalidated: keep the session, drop to RSSI now,
            // retry UWB after bounded backoff.
            fallbackToRSSI()
            scheduleUWBRetry()
        }
    }

    private func fallbackToRSSI() {
        measurement = Measurement(timestamp: Date(), method: .bleRssi,
                                  distanceMeters: nil, horizontalAngleRadians: nil,
                                  proximity: rssiFilter.band)
        stateText = "UWB 신호 없음 · 근접도만 표시"
    }

    private func scheduleUWBRetry() {
        guard let selection, selection.method == .niPeer || selection.method == .uwbAppleInterop else { return }
        retryTask?.cancel()
        let delay = backoff.next()
        retryTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            guard let self, self.active, !Task.isCancelled else { return }
            // A retry uses a fresh attemptId (§12): the offerer re-offers, the
            // controlee awaits the peer's new offer / apple_config.
            await MainActor.run { self.beginAttempt() }
        }
    }
}

File: android/app/src/main/java/com/cellularchat/app/ranging/RangingCoordinator.kt
package com.cellularchat.app.ranging

import com.cellularchat.app.core.cbor.CborBytes
import com.cellularchat.app.core.cbor.CborInt
import com.cellularchat.app.core.cbor.CborMap
import com.cellularchat.app.core.cbor.cborMapOf
import com.cellularchat.app.core.protocol.CapabilitySet
import com.cellularchat.app.core.protocol.CapabilityTranscript
import com.cellularchat.app.core.protocol.RangingMethod
import com.cellularchat.app.core.protocol.RangingSelector
import com.cellularchat.app.core.protocol.SessionMsgType
import java.util.UUID

/**
 * Ranging selection + fallback ladder (PROTOCOL_V2.md §12,
 * IMPLEMENTATION_PLAN.md §8). Picks a method deterministically from both
 * CapabilitySets, drives the matching controller, and always keeps BLE RSSI as
 * the fallback. On UWB failure it degrades to proximity immediately and retries
 * UWB with bounded backoff — a ranging failure is never reported as a pairing
 * or authentication failure.
 *
 * The two UWB controllers are injected so this orchestration is exercised
 * without hardware; the RSSI path is fully pure.
 */
class RangingCoordinator(
    private val output: Output,
    private val rawUwb: RawUwbController? = null,
    private val androidOob: AndroidOobController? = null,
    private val filter: RssiProximityFilter = RssiProximityFilter(),
    private val backoff: BackoffSchedule = BackoffSchedule(),
) {
    interface Output {
        fun onDirection(measurement: Measurement)
        fun onDistance(measurement: Measurement)

        /** An RSSI-derived proximity band, with its advisory trend (Feature C). */
        fun onProximity(band: ProximityBand, trend: RssiTrend, confidence: TrendConfidence)
        fun onRangingUnavailable(detail: String)
        fun onSignalLost()

        /** The platform actually started ranging with [technology] (a
         * RangingManager technology constant) so the UI can show it (§8/§12). */
        fun onTechnology(technology: Int)

        /** Emit an authenticated session message (apple_config/oob_data/…). */
        fun sendSessionMessage(msgType: Long, body: CborMap)

        /** Schedule [action] after [delayMillis]; injectable for tests. */
        fun scheduleRetry(delayMillis: Long, action: () -> Unit)

        /**
         * A §14 capability-transcript violation observed in a ranging message
         * (Feature B.2.2–B.2.4): the session layer must disconnect with
         * `capabilityMismatch` and fail the logical session — never a fallback.
         */
        fun onCapabilityMismatch() {}
    }

    private var method: Int = RangingMethod.BLE_RSSI
    private var attemptId: Long = 0
    private var peerUuid: UUID? = null
    private var oobInitiator: Boolean = false
    private var stopped = false

    // The two CapabilitySets bound to this logical session (Feature B); the
    // mutually-supported-method predicate is computed from them.
    private var localCaps: CapabilitySet? = null
    private var peerCaps: CapabilitySet? = null
    // Offered method per attemptId, to catch an accept whose method diverges (B.2.3).
    private val offeredMethods = mutableMapOf<Long, Int>()

    /** True while the RSSI band path is driving the UI (the fallback or the
     * selected method); cleared once a UWB attempt produces a fresh sample. */
    private var proximityActive = false

    // Background-ranging gating (§8): non-UWB ranging is foreground-only, and
    // background UWB runs only when the local device reports support. When the
    // gate says stop, ranging is paused (controllers stopped) while the
    // authenticated session stays up; it resumes on return to the foreground.
    private var foreground = true
    private var backgroundRangingSupported = false
    private var paused = false

    /** Selects the method for this pair (identical result on both phones, §12). */
    fun select(local: CapabilitySet, peer: CapabilitySet): Int {
        localCaps = local
        peerCaps = peer
        method = RangingSelector.select(local, peer).method
        backgroundRangingSupported = local.backgroundRanging
        return method
    }

    /**
     * Reports the app/service foreground state (§8). Backgrounding pauses ranging
     * unless the gate permits it (background UWB with device support); returning
     * to the foreground resumes from a fresh attempt. The Noise session is never
     * torn down here — only ranging work is gated.
     */
    fun setForeground(value: Boolean) {
        if (foreground == value) return
        foreground = value
        if (stopped) return
        if (value) {
            if (paused) {
                paused = false
                filter.reset()
                backoff.reset()
                beginAttempt()
            }
        } else if (!BackgroundRangingGate.shouldRange(false, method, backgroundRangingSupported)) {
            pauseRanging()
        }
    }

    private fun pauseRanging() {
        if (paused) return
        paused = true
        proximityActive = false
        runCatching { rawUwb?.stop() }
        runCatching { androidOob?.stop() }
        filter.reset()
    }

    /**
     * Begins ranging for the chosen method. [peerUuid] is this device's install
     * id of the peer (RangingDevice UUID); [oobInitiator] is the OOB role.
     */
    fun start(peerUuid: UUID, oobInitiator: Boolean) {
        this.peerUuid = peerUuid
        this.oobInitiator = oobInitiator
        stopped = false
        paused = false
        backoff.reset()
        filter.reset()
        beginAttempt()
    }

    private fun beginAttempt() {
        if (stopped) return
        attemptId += 1
        when (method) {
            RangingMethod.UWB_APPLE_INTEROP -> startAppleInterop()
            RangingMethod.UWB_ANDROID_OOB -> startAndroidOob()
            // BLE_RSSI (or NI_PEER, impossible on Android) is the proximity path.
            else -> { proximityActive = true; emitProximityIfAny() }
        }
    }

    private fun startAppleInterop() {
        val controller = rawUwb ?: return useRssiFallback("이 기기는 UWB를 지원하지 않습니다.")
        val config = controller.buildAccessoryConfig() ?: return useRssiFallback("UWB를 사용할 수 없습니다.")
        output.sendSessionMessage(
            SessionMsgType.APPLE_CONFIG,
            cborMapOf(1L to CborInt(attemptId), 2L to CborBytes(config)),
        )
        // Awaiting apple_shareable; RSSI keeps the UI alive until UWB locks.
        emitProximityIfAny()
    }

    private fun startAndroidOob() {
        val controller = androidOob ?: return useRssiFallback("이 기기는 UWB를 지원하지 않습니다.")
        val uuid = peerUuid ?: return useRssiFallback("상대 식별자가 없습니다.")
        controller.start(oobInitiator, uuid)
        emitProximityIfAny()
    }

    /**
     * Routes an inbound ranging session message. Before adopting a negotiation
     * message it enforces the capability transcript (Feature B.2.2–B.2.4): a
     * method (or implied method) outside the mutually-supported set, or an accept
     * that diverges from its offer, raises `capabilityMismatch` up to the session
     * layer instead of being silently dropped.
     */
    fun onSessionMessage(msgType: Long, body: CborMap) {
        when (msgType) {
            SessionMsgType.RANGING_OFFER -> {
                val method = methodField(body) ?: return // no method: legacy no-op (§10 idempotency)
                if (!methodSupported(method)) return raiseCapabilityMismatch()
                (body[1L] as? CborInt)?.value?.let { offeredMethods[it] = method }
            }
            SessionMsgType.RANGING_ACCEPT -> {
                val method = methodField(body) ?: return
                if (!methodSupported(method)) return raiseCapabilityMismatch()
                val attempt = (body[1L] as? CborInt)?.value
                val offered = attempt?.let { offeredMethods[it] }
                if (offered != null && offered != method) return raiseCapabilityMismatch()
            }
            SessionMsgType.APPLE_CONFIG -> {
                // Implicit offer for uwb_apple_interop (B.2.4).
                if (!methodSupported(RangingMethod.UWB_APPLE_INTEROP)) return raiseCapabilityMismatch()
            }
            SessionMsgType.NI_TOKEN -> {
                // Implicit offer for ni_peer (B.2.4).
                if (!methodSupported(RangingMethod.NI_PEER)) return raiseCapabilityMismatch()
            }
            SessionMsgType.APPLE_SHAREABLE -> {
                val controller = rawUwb ?: return
                val uuid = peerUuid ?: return
                val data = (body[2L] as? CborBytes)?.value ?: return
                controller.startFromShareable(data, uuid)
            }
            SessionMsgType.OOB_DATA -> {
                // Implicit offer for uwb_android_oob (B.2.4): verify before adopting.
                if (!methodSupported(RangingMethod.UWB_ANDROID_OOB)) return raiseCapabilityMismatch()
                val data = (body[2L] as? CborBytes)?.value ?: return
                androidOob?.onOobData(data)
            }
        }
    }

    private fun methodField(body: CborMap): Int? = (body[2L] as? CborInt)?.value?.toInt()

    private fun raiseCapabilityMismatch() = output.onCapabilityMismatch()

    /** The mutually-supported-method predicate over both bound CapabilitySets (B.2.2). */
    private fun methodSupported(candidate: Int): Boolean {
        val local = localCaps ?: return true // pre-select: cannot evaluate; never false-positive
        val peer = peerCaps ?: return true
        return CapabilityTranscript.methodSupported(local, peer, candidate)
    }

    /** Feeds one raw BLE RSSI reading (dBm); drives the UI only on the band path. */
    fun feedRssi(rssiDb: Int) {
        if (stopped) return
        val band = filter.update(rssiDb)
        // RSSI is non-UWB, so it is foreground-only (§8). The advisory trend
        // (Feature C) accompanies every RSSI-derived proximity value.
        if (proximityActive && foreground) output.onProximity(band, filter.trend, filter.trendConfidence)
    }

    private fun emitProximityIfAny() {
        val band = filter.current()
        if (proximityActive && foreground && band != ProximityBand.UNKNOWN) {
            output.onProximity(band, filter.trend, filter.trendConfidence)
        }
    }

    private fun useRssiFallback(detail: String) {
        // Keep the authenticated link; degrade to proximity and retry UWB later.
        proximityActive = true
        output.onRangingUnavailable(detail)
        emitProximityIfAny()
        if (method != RangingMethod.BLE_RSSI) scheduleUwbRetry()
    }

    private fun scheduleUwbRetry() {
        if (stopped) return
        val delay = backoff.nextDelayMillis()
        output.scheduleRetry(delay) { if (!stopped) beginAttempt() }
    }

    /** Callbacks a UWB controller reports through. */
    val uwbCallbacks: RawUwbController.Callbacks = object : RawUwbController.Callbacks {
        override fun onStarted(technology: Int) {
            if (stopped) return
            // Surface the technology the platform ACTUALLY selected, not the
            // requested method (§8/§12).
            output.onTechnology(technology)
        }

        override fun onMeasurement(distanceMeters: Double?, azimuthDegrees: Double?, elevationDegrees: Double?) {
            if (stopped) return
            backoff.reset()
            when {
                azimuthDegrees != null -> {
                    proximityActive = false
                    output.onDirection(Measurement(method, distanceMeters, azimuthDegrees, elevationDegrees))
                }
                distanceMeters != null -> {
                    proximityActive = false
                    output.onDistance(Measurement(method, distanceMeters))
                }
                // Neither yet: hold with proximity so nothing stale is shown.
                else -> emitProximityIfAny()
            }
        }

        override fun onError(rangingErrorCode: Int, detail: String) {
            if (stopped) return
            useRssiFallback(detail)
        }

        override fun onStopped() {
            if (stopped) return
            // Feature C.8: clear the trend on signal loss so nothing stale shows.
            filter.reset()
            output.onSignalLost()
            scheduleUwbRetry()
        }
    }

    fun stop() {
        stopped = true
        runCatching { rawUwb?.stop() }
        runCatching { androidOob?.stop() }
        filter.reset()
    }

    fun close() {
        stop()
        runCatching { rawUwb?.close() }
        runCatching { androidOob?.close() }
    }
}
